src/gcp/cloudsql/connect.ts

```diff
@@ -105,82 +106,102 @@ export async function execute(
-  const { user, mode } = toDatabaseUser(account);
-  await cloudSqlAdminClient.createUser(projectId, instanceId, mode, user);
-
-  // 3. Connect to the DB as the temporary user and run the necessary grants
-  // TODO: I think we're missing something here, sometimes backend can't see the tables.
-  const grants = [
-    `do
-      $$
-      begin
-        if not exists (select FROM pg_catalog.pg_roles
-          WHERE  rolname = '${firebaseowner(databaseId)}') then
-          CREATE ROLE "${firebaseowner(databaseId)}" WITH ADMIN "${setupUser}";
-        end if;
-      end
-      $$
-      ;`,
-    `GRANT ALL PRIVILEGES ON DATABASE "${databaseId}" TO "${firebaseowner(databaseId)}"`,
-    `GRANT cloudsqlsuperuser TO "${firebaseowner(databaseId)}"`,
-    `GRANT "${firebaseowner(databaseId)}" TO "${setupUser}"`,
-    `GRANT "${firebaseowner(databaseId)}" TO "${user}"`,
-    `ALTER SCHEMA public OWNER TO "${firebaseowner(databaseId)}"`,
-    `GRANT USAGE ON SCHEMA "public" TO PUBLIC`,
-    `GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" TO PUBLIC`,
-    `GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" TO PUBLIC`,
-  ];
-  await execute(grants, {
+  return await execute([`SET ROLE = cloudsqlsuperuser`, ...cmds], {
     projectId,
     instanceId,
     databaseId,
-    username: setupUser,
+    username: superuser,
     password: temporaryPassword,
-    silent: true,
+    silent: silent,
   });
-  return user;
 }
 
-export function firebaseowner(databaseId: string) {
-  return `firebaseowner_${databaseId}_public`;
+export async function getIAMUser(options: Options): Promise<{ user: string; mode: UserType }> {
+  const account = await requireAuth(options);
+  if (!account) {
+    throw new FirebaseError(
+      "No account to set up! Run `firebase login` or set Application Default Credentials",
+    );
+  }
+
+  return toDatabaseUser(account);
+}
+
+// setupIAMUser sets up the current user identity to connect to CloudSQL.
+// Steps:
+// 2. Create an IAM user for the current identity
+// 3. Connect to the DB as the temporary user and run the necessary grants
+// 4. Deletes the temporary user
+export async function setupIAMUser(
+  instanceId: string,
+  databaseId: string,
+  options: Options,
+): Promise<string> {
+  // TODO: Is there a good way to short circuit this by checking if the IAM user exists and has the appropriate role first?
+  const projectId = needProjectId(options);
+
+  // 0. Get the current identity
+  const { user, mode } = await getIAMUser(options);
+
+  // 1. Create an IAM user for the current identity.
+  await cloudSqlAdminClient.createUser(projectId, instanceId, mode, user);
+
+  // 2. Setup FDC required SQL roles and permissions.
+  await setupSQLPermissions(instanceId, databaseId, options, true);
+
+  // 3. Grant firebaseowner role to the IAM user.
+  const grants = [`GRANT "${firebaseowner(databaseId)}" TO "${user}"`];
+
+  await executeSqlCmdsAsSuperUser(options, instanceId, databaseId, grants, true);
```

Comment:
nit: I was looking for GRANT current user owner in `setupSQLPermissions`.

Re-naming suggestions to the line easier: What about?
- `setupIAMUser` just upserts two IAM user (user and P4SA)
- `setupSQLPermissions` just deals with SQL permissions. The last grant is also included in the big setup SQL array.

The caller calls both.

Reply:
Responsed in another thread. Willing to change if you still think it's better to combine them, the biggest reason is that I wanted `setupSQLPermissions` to be what runs in `dataconnect:sql:setup`.

Reply:
Moved p4sa creation and grants to setupIamUsers after our convo